Table.fromRValuesStream, with column awareness

edmundnoble opened the pull request.

rintcius commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala:
```diff
@@ -1894,16 +1894,95 @@ object Slice {
+            // println("we have too many columns, cutting slice early")
+            (new Slice {
+              val size = rows
+              val columns = acc
```
Continuing discussion of previous PR..

> `I think we don't want to make slices smaller than minSliceRows.`

Not by default, but if we have a column overflow then we can end up with smaller slices, even with just 1 row. Looks to me like it's worth down-sizing the array allocations in that scenario.

rintcius commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala:
```diff
@@ -1894,16 +1894,95 @@ object Slice {
+    fromRValues(values.map(RValue.fromJValueRaw))
 
+  // don't use this anymore. It doesn't limit the slice size properly,
+  // unlike allFromRValues and fromRValuesStep.
```
Can't we get rid of this function yet?

edmundnoble replied:
I was told we should leave this for the old API.

alissapajer replied:
Once the old API is gone, we should be able to move this function into a `test/` dir.

rintcius commented on yggdrasil/src/test/scala/quasar/yggdrasil/table/SliceSpec.scala:
```diff
@@ -71,6 +72,221 @@ class SliceSpec extends Specification with ScalaCheck {
+    "construct slices from pathological example" in {
+      // this example is pathological because when dealing with it there's a difference between cutting off
+      // a slice only when the column limit is reached and deciding beforehand whether adding a value would
+      // exceed the column limit.
```
looks like a nice test, but I don't get your comment, can you elaborate a bit?

edmundnoble replied:
The difference here is between two algorithms for deciding when the column limit should force us to move to the next slice:
a) I inspected an RValue and decided it would put us above the column limit, so we'll leave it for the next slice, and
b) We always add values if we're currently below the column limit.

I implemented b) first, but it failed one of your tests (the random one), and this was the sample which did it.

rintcius replied:
ok, thanks for clarifying @edmundnoble

alissapajer replied:
Could you instead say something like:

"If we add an `RValue` to the current `Slice` without checking if its addition will exceed the column limit, it's possible that we will indeed exceed the column limit. This test case hit that bug, so we include it here."

The comment as it is doesn't say what the bug was, which is why I think it was a little unclear.

edmundnoble replied:
Will do.

rintcius commented on yggdrasil/src/test/scala/quasar/yggdrasil/table/SliceSpec.scala:
```diff
@@ -71,6 +72,221 @@ class SliceSpec extends Specification with ScalaCheck {
+      "maxSliceRows < nrRows" >> testFromRValuesMaxSliceRowsOverflow(v)
+      "maxSliceRows = 1" >> testFromRValuesMaxSliceRows1(v)
+      "maxSliceColumns = nrColumns of biggest value" >> testFromRValuesMaxSliceColumnsEqualsBiggestValue(v)
+      // "maxSliceColumns < nrColumns of biggest value" >> testFromRValuesMaxSliceColumnsLowerThanBiggestValue(v)
```
why is this commented out?

edmundnoble replied:
Good point.

djspiewak commented on foundation/src/main/scala/quasar/contrib/fs2/stream.scala:
```diff
@@ -16,8 +16,9 @@
-import slamdata.Predef.{Throwable, Unit}
+import slamdata.Predef.Throwable
 
+import scala.Unit
```
It is.

wemrysi commented:
@edmundnoble Did downgrading `cats-effect`/`fs2` actually have an effect on performance, I thought we determined the issue was still present in the `fs2-0.10.x` series.

wemrysi commented:
@edmundnoble Oh, nevermind, I was thrown off track by one of the commit messages, looks like you undid all the reverting.

edmundnoble commented:
@wemrysi I undid it all, yeah. I decided to leave that stuff to later.

edmundnoble commented on it/src/main/resources/tests/temporal/dateFilter.test:
```diff
@@ -1,10 +1,6 @@
-    "backends": {
-        "lwc_local": "pending",
-        "mimir": "pending"
-    },
```
It was making `fromRValues` not go through `JValue` first, because `JValue` destroys date/time values.

djspiewak commented on mimir/src/main/scala/quasar/mimir/evaluate/FederatedShiftedReadPlanner.scala:
```diff
@@ -44,7 +39,7 @@ import scalaz._, Scalaz._
 final class FederatedShiftedReadPlanner[
     T[_[_]]: BirecursiveT: EqualT: ShowT,
-    F[_]: LiftIO: Monad: PlannerErrorME: MonadFinalizers[?[_], IO]](
+    F[_]: LiftIO: Monad: PlannerErrorME: MonadTell_[?[_], DList[IO[Unit]]]](
```
I think there's also a type alias, right?

rintcius replied:
@djspiewak those aliases are removed in this PR, see https://github.com/slamdata/quasar/pull/3644/files#diff-2ee44a217677c1a18c6b494972360fdcL30 - personally I am not that opinionated about these aliases, just a slight preference to keep them in, but happy to go with other opinions

alissapajer replied:
I'd prefer to keep the aliases in; I think the name helps clarify the type since it's a type that includes `Unit`.

edmundnoble replied:
The aliases are causing compiler errors because dealiasing is somehow interacting with implicit search, I think.

edmundnoble replied:
I also don't know where to put them anymore; they're used in yggdrasil now, not just mimir, and there's already an yggdrasil package object but it's just in the test project (and you can't have two). I could put it in foundation but it's not used anywhere else.

alissapajer commented:
[ch71]

alissapajer commented on yggdrasil/src/main/scala/quasar/yggdrasil/TableModule.scala:
```diff
@@ -35,6 +36,7 @@ import java.time.{LocalDate, LocalDateTime, LocalTime, OffsetDateTime, OffsetTim
 
 // TODO: define better upper/lower bound methods, better comparisons,
 // better names, better everything!
+// TODO: investigate adding columns
```
What does this comment mean?

edmundnoble replied:
I meant specifically adding them to TableSize, I've tried now to make that more explicit.

alissapajer commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/ArrayColumn.scala:
```diff
@@ -33,9 +33,25 @@ trait DefinedAtIndex {
 
-class ArrayHomogeneousArrayColumn[@specialized(Boolean, Long, Double) A](val defined: BitSet, values: Array[Array[A]])(val tpe: CArrayType[A])
+object ArrayColumn {
+  def resizeArray[A](arr: Array[A], size: Int)(implicit A: ClassTag[A]): Array[A] = {
```
This will lose values from `arr` if a `size` is chosen that is smaller than the number of values in `arr`. Is that the intention?

edmundnoble replied:
Fixed, now does nothing in that case.

alissapajer commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/ArrayColumn.scala:
```diff
@@ -33,9 +33,25 @@ trait DefinedAtIndex {
+  def resizeBitSet(bs: BitSet, size: Int): BitSet = {
+    val arr = new Array[Long](size)
+    // 64 bits per long, 2^6 == 64
+    System.arraycopy(bs.getBits(), 0, arr, 0, size >> 6)
```
Say `size` is `200`. Then `size >> 6` is `3`. This means we'd copy over three `Long` values. 

What's interesting here is that we are copying over the bits in `Long`-sized (64-bit sized) chunks. So, when `size` is `200` meaning (unless I've misunderstood) that we want to copy over `200` bits, we'd actually only copy over `64 * 3 = 192` bits.

alissapajer replied:
Also, we're creating an `Array[Long]` of `size` but then only copying over `size >> 6` elements.

edmundnoble replied:
Ooh, missing data, with the exception fixed by incorrect array sizing. This is starting to become a pattern.

edmundnoble replied:
So really this function will only be called for powers of two, for which this is accurate. However, in case that isn't the case, I've made it round up when necessary.

edmundnoble replied:
All fixed now, even another edge-case I discovered during testing.

alissapajer commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala:
```diff
@@ -1894,16 +1894,95 @@ object Slice {
+    def tail: ArraySliced[A] = ArraySliced(arr, start + 1, size - 1)
+  }
+
+  def fromRValuesStep(values: ArraySliced[RValue], maxRows: Int, maxColumns: Int): (Slice, ArraySliced[RValue]) = {
```
This is production code, so we should have some unit tests for this function specifically. The logic is gnarly, and we're only hitting it in the `lwc_local` integration tests. We should make sure to test the edge cases.

alissapajer replied:
This function reminded me of this poem.
```
Parsley
Is gharsley.
--Ogden Nash
```

edmundnoble replied:
I'll definitely write these.

alissapajer commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala:
```diff
@@ -1894,16 +1894,95 @@ object Slice {
+    val maxRowsC = maxRows.getOrElse(Config.maxSliceRows)
+    val maxColumnsC = maxColumns.getOrElse(Config.maxSliceColumns)
+    // println(s"maxRows: $maxRowsC, maxCols: $maxColumnsC")
+    def rec(next: Slice.ArraySliced[RValue], values: fs2.Stream[IO, RValue]): fs2.Pull[IO, Slice, Unit] =
```
Can we make this `@tailrec`?

alissapajer commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala:
```diff
@@ -1894,16 +1894,95 @@ object Slice {
+  }
+
+  def fromRValuesStep(values: ArraySliced[RValue], maxRows: Int, maxColumns: Int): (Slice, ArraySliced[RValue]) = {
+    def inner(next: ArraySliced[RValue], rows: Int, cols: Int, acc: Map[ColumnRef, ArrayColumn[_]], _size: Int): (Slice, ArraySliced[RValue]) = {
```
Same, can this be `@tailrec`?

alissapajer replied:
@djspiewak (or anyone) hm, are there performance concerns we need to think about regarding tailrec-ness and doing computations on the stack as we recurse?

djspiewak replied:
`@tailrec` is considerably faster than regular recursion when you're talking about the hot path. It doesn't have to allocate new stack frames (which is much faster than heap allocation, but still a cost) and doesn't have to re-check polymorphism on the recursion, etc. It literally turns into a `goto`.

edmundnoble replied:
I've now marked this `@tailrec`, but just so you know @alissapajer the compiler would optimize the tail-recursion even without the annotation.

alissapajer commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala:
```diff
@@ -1894,16 +1894,95 @@ object Slice {
+      } else {
+        // println(s"extracting slice from data with size ${next.size}")
+        val (nextSlice, remainingData) = Slice.fromRValuesStep(next, maxRowsC, maxColumnsC)
+        fs2.Pull.output1(nextSlice) >> rec(remainingData, values)
```
I haven't followed all the logic through completely yet, but here's one potential concern: Let's say we take a chunk from the `Stream` and this chunk has 1000 rows, the first 998 of which we turn into a `Slice`. Using this algorithm, won't those remaining two rows turn into a `Slice` of size two? 

If so, then we'd have to `canonicalize` in order to get a `Table` with correctly-sized `Slice`s. We have to pay the cost at some point of reassociating slice boundaries (whether those be actual `Slice`s or chunks coming in from `fs2.Stream`). I'm not sure yet if that cost is better paid within this function, or calling `canonicalize` is okay.

edmundnoble replied:
I actually was under the impression we didn't want to do this right now, from what we went through in the meeting room. However, that's not incredibly difficult to implement.

@alissapajer I believe paying the cost in this function is significantly better, because otherwise we would be allocating the slices twice.

edmundnoble commented:
Much of the code in `fromRValuesStep` is completely incorrect. While working on performance tests for mimir, I found a hint that this function was implemented incorrectly when it crashed with OOM when I was producing large columns. Here's a list of the things that were wrong:

- `updateRefs` was only updating the index at `maxRows`. This meant all rows were empty but one.
- We weren't keeping track of the row counts per column separately. This means that having three columns each with 10 rows would lead to the method thinking you had 30 rows in the last column.
- Rows were added *after* checking that the *current* row count didn't exceed `_size` or `maxRows`, leading to array indexing exceptions. This wasn't noticeable until the previous bug was fixed.

Here's a list of the things that were noticed that were performance bugs:
- Minimum slice size of 64 is unnecessarily pessimistic for flat data. We can find out how much of the data is scalar and use that to make a larger estimate when it's feasible.
- ArrayBoolColumn didn't have an `empty(size: Int)` constructor, leading to runtime resizing of the `BitSet`.
- Tuples (`->`) in `updateRefs`, `foldLeft` in `updateRefs`.

And for @alissapajer:
- `ArrayColumn.resize` was made to return the same column instead of making the column smaller.

I've also found a bug, for which I've left a comment:
```scala
          // this shouldn't work. `updateRefs` mutates the underlying arrays of `acc`,
          // so just "throwing away" `newAcc` later on when we decide we don't
          // want it anymore is not possible.
          val newAcc = updateRefs(flattened, acc, rowsInCol, _size)
          val newCols = newAcc.size
          if (newCols > maxColumns && cols > 1) {
            (new Slice {
              val size = rows
              val columns = acc
            }, next)
          } else {
            inner(next.tail, newRowsInCol, newRows, newCols, newAcc, _size)
          }
        }
```

edmundnoble commented on foundation/src/main/scala/quasar/contrib/fs2/convert.scala:
```diff
@@ -23,7 +23,7 @@ import java.util.Iterator
 import scala.util.{Either, Left, Right}
 
-import cats.effect.{Concurrent, Sync}
+import cats.effect.{ConcurrentEffect, Sync}
```
For `runAsync` in `chunkQ`, it seems.

wemrysi replied:
I think we’re using `Concurrent#start` in `master`, that’d obviate the need for `ConcurrentEffect`.

edmundnoble replied:
Fixed.

djspiewak commented:
Test failures look legit

edmundnoble commented:
Just to repeat what I said in a tiny, probably now hidden comment to Alissa: please don't merge this without unit tests, which I am working on today.

edmundnoble commented:
Unit tests are in; let me know if I missed any, but otherwise I believe I've addressed all of the feedback, though @wemrysi if you have any opinions on the finalizer type alias stuff I'd like to push those into this PR, because it would only reduce the number of lines changed.

alissapajer commented on blueeyes/src/main/java/quasar/precog/BitSet.java:
```diff
@@ -490,6 +490,21 @@ public boolean equals(Object obj) {
         return true;
     }
 
+    public String toString() {
```
Why was this added?

edmundnoble replied:
Testing.

alissapajer commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/ArrayColumn.scala:
```diff
@@ -69,6 +107,28 @@ class ArrayBoolColumn(val defined: BitSet, val values: BitSet) extends ArrayColu
+  override def equals(other: Any): Boolean = {
+    if (other.isInstanceOf[ArrayBoolColumn]) {
+      val otherC = other.asInstanceOf[ArrayBoolColumn]
+      // this is incorrect, we should filter by definedness first.
```
Can we make this do the right thing in this PR? Someone might use this function expecting it to do the right thing without reading this comment.

Also, I was curious, were the `equals` methods you implemented here used just for testing, or are they used elsewhere too?

edmundnoble replied:
Fixed this in the subsequent commit. They're just for testing.

alissapajer commented:
This failed to compile. Also, merge master to get the temporary fix (I hope) for the SOE!

alissapajer commented:
The travis build was finding a stale cache so I deleted the cache for this PR build and restarted the build :crossed_fingers:

alissapajer commented:
@edmundnoble some tests are fixed now in the `lwc_local` travis build.

alissapajer commented:
@rintcius @wemrysi since you've both requested changes, I won't merge this until each of you approve.

alissapajer commented on yggdrasil/src/main/scala/quasar/yggdrasil/vfs/vfs.scala:
```diff
@@ -49,7 +49,9 @@ import shims.monoidToCats
-import scala.util.Either
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Promise
+import scala.util.{Failure, Success}
```
Are these imports used?

wemrysi commented:
@edmundnoble Regarding the `MonadFinalizer` alias, all I have to recommend is the `yggdrasil` package object, which would mean refactoring the tests to put their constructs somewhere else.

alissapajer commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala:
```diff
@@ -1784,126 +1792,284 @@ object Slice {
+  ): (Slice, ArraySliced[RValue]) = {
+    @tailrec
+    def inner(
+      next: ArraySliced[RValue], rows: Int, cols: Int,
```
It doesn't look like `cols` is used anywhere.

alissapajer commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala:
```diff
@@ -1784,126 +1792,284 @@ object Slice {
+    @tailrec
+    def inner(
+      next: ArraySliced[RValue], rows: Int, cols: Int,
+      acc: Map[ColumnRef, ArrayColumn[_]], _size: Int
```
What do you think of renaming `_size` to `allocatedColSize`? There are a few different "sizes" here so I think a more specific name would help readability.

alissapajer commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala:
```diff
@@ -1784,126 +1792,284 @@ object Slice {
+          val flattened = next.head.flattenWithPath
+          val newAcc = updateRefs(flattened, acc, rows, _size)
+          val newCols = newAcc.size
+          val newRows = rows + 1
```
This is only used in the `else` clause on line 2001 so it can be within that context.

edmundnoble replied:
:+1:

edmundnoble commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala:
```diff
@@ -1784,126 +1792,284 @@ object Slice {
+          // by two.
+          // println("we're resizing the columns because we have too many rows")
+          val newSize = _size * 2
+          inner(next, rows, cols, acc.mapValues { _.resize(newSize) }, newSize)
```
This is very valuable to know, thank you.

edmundnoble commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala:
```diff
@@ -1784,126 +1792,284 @@ object Slice {
+      // get before we have too many columns.
+      var ctr = values.start
+      var r = values.head
+      while (ctr < values.arr.length - 1 && ctr < (values.start + maxRows) && RValue.toCValue(r).nonEmpty) {
```
Good point on always receiving scalars, but do we know for sure this will always be the case? Especially for LWC's.

alissapajer replied:
I view this as an optimization for the specific case of non-nested data. It won't always be the case that our customers send nested data, but in my understanding of the product, it will be rare for customers to send raw values not nested in any structure since we're a solution for "heterogeneous messy data". And so we'd be optimizing for the scalar case at the cost of the non-scalar case when really we should be doing the opposite. 

@djspiewak if my assumptions above are incorrect please advise. :)

edmundnoble replied:
I don't actually believe this is a significant cost to pay, is the thing. Type checks like this are fairly cheap, and they allow us to do things orders of magnitude faster when it's possible. But I have benchmarks now that can provide real evidence, and I'd like to revisit this with that benchmark PR.

alissapajer replied:
This is an optimization for a corner case in customer data. Sometimes optimizations like this are worth it. But for now it's extra logic that we need to maintain, reason about, debug, and test. And so unless we know it's going to be giving us real benefit, I think it should be omitted.

I'm also not convinced that we would see a performance benefit in the case of data that is entirely comprised of scalars, since we'll be looping through the `while` loop for every single value. It definitely could be faster, but we shouldn't be adding anything like this without benchmarks.

Could we omit it in this PR and then benchmark it? Then we can have a discussion with the benchmarks in hand. I think it's going to be hard to discuss this further without benchmark data.

edmundnoble replied:
I actually just checked with benchmarks, and this doesn't make anything faster. I removed it in a follow-up PR.

alissapajer commented:
@rintcius any other review to add?

edmundnoble commented on it/src/main/resources/tests/orderByIfUndefined.test:
```diff
@@ -2,7 +2,7 @@
     "backends": {},
     "data": "smallZips.data",
-    "query": "select city ?? \"foo\" as name from smallZips order by city ?? \"foo\"",
+    "query": "select city ?? \"aaa\" as name from smallZips order by city ?? \"aaa\"",
```
Well not exactly need; however, in the case that any `city` actually *was* `undefined`, it would not show up in the initial results (and the predicate is only on initial results) because `foo` is alphabetically so far down.

edmundnoble commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/ArrayColumn.scala:
```diff
@@ -69,6 +100,27 @@ class ArrayBoolColumn(val defined: BitSet, val values: BitSet) extends ArrayColu
+    } else {
+      false
+    }
+  }
```
This is only for testing.

djspiewak replied:
Still worth updating `hashCode`. Or perhaps defining a method *other* than `equals`, since it's just for testing.

alissapajer replied:
If this is only for testing we can name it `testEquals`.

edmundnoble commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/ArrayColumn.scala:
```diff
@@ -69,6 +100,27 @@ class ArrayBoolColumn(val defined: BitSet, val values: BitSet) extends ArrayColu
+    new ArrayBoolColumn(newDefined, newValues)
+  }
+
+  override def equals(other: Any): Boolean = {
```
Similarly, this is only for testing.

edmundnoble commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/ArrayColumn.scala:
```diff
@@ -93,6 +145,24 @@ class ArrayLongColumn(val defined: BitSet, val values: Array[Long]) extends Arra
+    if (other.isInstanceOf[ArrayLongColumn]) {
+      val otherC = other.asInstanceOf[ArrayLongColumn]
+      ArrayColumn.filterDefined(defined, values).deep ==
+        ArrayColumn.filterDefined(otherC.defined, otherC.values).deep
```
Ditto. I don't think there's any reason to use this in production; if we did, there are many other optimizations we could perform.

rintcius replied:
At the moment not, but we don't know for the future and if you don't know the history of the codebase then things like this gets missed easily.

Anyway, I'm ok with renaming `equals` to something like `testEquals` (and avoid any changes to `hashCode` and the impl of `equals`)

rintcius replied:
I think it is a bit better if we move `equals` under the test source root instead, but then it will be a bit more work. I'm also fine with just renaming.

edmundnoble commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala:
```diff
@@ -73,6 +73,14 @@ trait Slice { source =>
+      size == otherSlice.size && columns == otherSlice.columns
+    } else {
+      false
+    }
```
Ditto.

alissapajer replied:
If this is just used for testing we should move it into the test directory.

edmundnoble commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala:
```diff
@@ -1784,126 +1792,283 @@ object Slice {
+  def updateRefs(rv: List[(CPath, CValue)], into: Map[ColumnRef, ArrayColumn[_]], sliceIndex: Int, sliceSize: Int): Map[ColumnRef, ArrayColumn[_]] = {
+    var acc = into
+    var cursor = rv
+    while (cursor.isInstanceOf[::[(CPath, CValue)]]) {
```
It's faster, I've benchmarked this for scalaz previously.

rintcius replied:
ok, can you add a comment for that?

edmundnoble replied:
Oh compared to checking `Nil`, not pattern matching. They should be the same.

djspiewak replied:
It'll be slightly faster than checking `eq Nil` since it won't have to do a field lookup. It's negligible though. Naturally it's faster than pattern matching. You basically shouldn't use anything other than `@switchcase` pattern matching in the hot path.

edmundnoble commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala:
```diff
@@ -1784,126 +1792,283 @@ object Slice {
+        arr.deep == otherA.arr.deep && start == otherA.start && size == otherA.size
+      } else {
+        false
+      }
```
Ditto.

alissapajer replied:
Same, is this just used for testing?

djspiewak commented:
@edmundnoble Branch is now conflicted.

edmundnoble commented:
I *believe* this is ready for merge after Appveyor, if @alissapajer and @rintcius have no more feedback.

I removed the premature optimization for slice size, and moved the `equals` stuff into `SliceSpec`. Also fixed unused imports.

rintcius commented:
I have approved. There's just a couple of `equals` that are still in prod code as @alissapajer already pointed out. I'll leave that to her.

edmundnoble commented:
Good point @rintcius, moving that now.
